core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1158,9 +1158,14 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
     @throws(classOf[Exception])
     def handleNewSession() {
       info("ZK expired; shut down all controller components and try to re-elect")
-      onControllerResignation()
-      inLock(controllerContext.controllerLock) {
-        controllerElector.elect
+      if (controllerElector.getControllerID() != config.brokerId) {
+        onControllerResignation()
+        inLock(controllerContext.controllerLock) {
+          controllerElector.elect
+        }
+      } else {
```

Comment:
@junrao
@ijuma 
After LeaderChangeListener.handleDataDeleted  to create the controller ephemeral data,  then in the KafkaController.handleNewSession  will first resign the controller thread and start to elect the leader, but in the :
   if(leaderId != -1) {
       debug("Broker %d has been elected as leader, so stopping the election process.".format(leaderId))
       return amILeader
    }
The leaderId is itself and it think there is other broker becoming the controller, however the controller ephemeral node is created by  itself,  and the controller context is clean and the controller thread is exit  in the onControllerResignation function before the controllerElector.elect